File: back/coordinates/AbstractQuadrant.php
<?php
include_once 'Quadrant.php';
abstract class AbstractQuadrant implements Quadrant {
    private int $x_sign;
    private int $y_sign;
    protected string $type;
    public function __construct(int $x_sign, int $y_sign, string $type) {
        $this->x_sign = sign($x_sign);
        $this->y_sign = sign($y_sign);
        $this->type = $type;
    }

    private function checkSign(float $x, float $y): bool {
        return ((sign($x)==$this->x_sign || sign($x)==0) && (sign($y)==$this->y_sign || sign($y)==0));
    }
    public function checkHit(float $x, float $y, float $r): bool {
        if ($this->checkSign($x,$y)) {
            return $this->calcHit($x,$y,$r);
        }
        return false;
    }
    protected abstract function calcHit(float $x, float $y, float $r): bool;
    public function jsonSerialize() {
        return [
            'type' => $this->type,
            'x_sign' => $this->x_sign,
            'y_sign' => $this->y_sign
        ];
    }
}
function sign($n) {
    return ($n > 0) - ($n < 0);
}

File: src/main/java/coordinates/AbstractQuadrant.java
package coordinates;

import com.fasterxml.jackson.annotation.JsonCreator;
import com.fasterxml.jackson.annotation.JsonProperty;

public abstract class AbstractQuadrant implements Quadrant {
    @JsonProperty
    private final int xSign;
    @JsonProperty
    private final int ySign;
    @JsonCreator
    public AbstractQuadrant(@JsonProperty("xSign") int xSign, @JsonProperty("ySign") int ySign) {
        this.xSign = xSign;
        this.ySign = ySign;
    }
    @Override
    public boolean checkHit(double x, double y, double r) {
        if (checkSign(x,y)) return calcHit(x,y,r);
        return false;
    }
    private boolean checkSign(double x, double y) {
        return (sign(x)==xSign || x==0) && (sign(y)==ySign || y==0);
    }
    protected abstract boolean calcHit(double x, double y, double r);

    protected int sign(int num) {
        return num>0 ? 1 : -1;
    }
    protected int sign(double num) {
        return num>0 ? 1 : -1;
    }

    @Override
    public String toString() {
        return "AbstractQuadrant{" +
                "xSign=" + xSign +
                ", ySign=" + ySign +
                '}';
    }
}
